src/objective/rank_obj.cc

```diff
@@ -64,12 +64,13 @@ class LambdaRankObj : public ObjFunction {
       std::vector< std::pair<bst_float, unsigned> > rec;
       bst_float sum_weights = 0;
       for (bst_omp_uint k = 0; k < ngroup; ++k) {
-        sum_weights += info.GetWeight(k);
+        sum_weights += info.GetWeight(gptr[k]);
```

Comment:
Thanks for pointing out the issue. I don't think we should fix the problem by changing the size of weight vector. This is because there have been several changes in the codebase that assumed `[size of weight vector] = [number of groups]`. See #4115, #4118, #4121.

I propose that we revise the metrics `auc` and `aucpr` instead.

Reply:
@hcho3 Thanks for the contextual information on all the relevant PRs. Given that there has been a lot of works on top of the assumption `[size of weight vector] = [number of groups]`  I agree that revising `auc` and `aucpr` behaviors is easier. I will rework the PR so that we change how `auc` and `aucpr` utilizes the group weights. 

It occurred to me that if I change the implementation in `rank_obj.cc`, I probably will need to change the python package as well as xgboost-4j wherever a ranker is to be trained. So I like your proposal.

Reply:
Great! One tricky spot would be to ensure that AUC will work for binary classification (where weight is per-instance) and for ranking (where weight is per-group).